ramips: add support for B-Link BL-R7628AE4

ysc3839 opened the pull request:
The model of this router is BL-WR4000. Only the PCB has the model name
"BL-R7628AE4" printed on it.

Specifications:
- MediaTek MT7628N/N (575 MHz)
- 32 MB RAM
- 4 MB Flash
- 802.11bgn radio
- 5x 10/100 Mbps Ethernet (1 WAN and 4 LAN)
- 7x LED (Power WLAN WAN LAN1-4)
- 1x Reset button

Flash instruction:

The only way to flash OpenWrt image in BL-R7628AE4 is to use
tftp recovery mode in U-Boot:

1. Configure PC with static IP 192.168.16.123/24 and tftp server.
2. Rename "openwrt-ramips-mt76x8-b-link_bl-r7628ae4-squashfs-sysupgrade.bin"
   to "uImage" and place it in tftp server directory.
3. Connect PC with the LAN port, press the reset button, power up
   the router and keep button pressed for around 6-7 seconds, until
   device starts downloading the file.
4. Router will download file from server, write it to flash and
   reboot.

Known issue:
1. Turning on/off wireless cause ethernet disconnect and then connect. `lede-17.01` branch do not have this issue.
2. WAN/LAN leds always blinking if the port is connected. Same issue in `lede-17.01` branch.

ysc3839 committed:
ramips: add support for B-Link BL-R7628AE4

The model of this router is BL-WR4000. Only the PCB has the model name
"BL-R7628AE4" printed on it.

Specifications:
- MediaTek MT7628N/N (575 MHz)
- 32 MB RAM
- 4 MB Flash
- 802.11bgn radio
- 5x 10/100 Mbps Ethernet (1 WAN and 4 LAN)
- 7x LED (Power WLAN WAN LAN1-4)
- 1x Reset button

Flash instruction:

The only way to flash OpenWrt image in BL-R7628AE4 is to use
tftp recovery mode in U-Boot:

1. Configure PC with static IP 192.168.16.123/24 and tftp server.
2. Rename "openwrt-ramips-mt76x8-b-link_bl-r7628ae4-squashfs-sysupgrade.bin"
   to "uImage" and place it in tftp server directory.
3. Connect PC with the LAN port, press the reset button, power up
   the router and keep button pressed for around 6-7 seconds, until
   device starts downloading the file.
4. Router will download file from server, write it to flash and
   reboot.

Signed-off-by: Richard Yu <yurichard3839@gmail.com>

mkresin commented on target/linux/ramips/base-files/lib/ramips.sh:
```diff
@@ -76,6 +76,9 @@ ramips_board_detect() {
 	*"BC2")
 		name="bc2"
 		;;
+	*"BL-R7628AE4")
```
No need to add your board to this file, since you're using the devicetree based board name.

mkresin commented on target/linux/ramips/dts/BL-R7628AE4.dts:
```diff
@@ -0,0 +1,98 @@
+&pinctrl {
+	state_default: pinctrl0 {
+		gpio {
+			ralink,group = "wled_an";
```
please use `ralink,group = "wdt", "wled_an";` here, as you're using the wdt gpio for the reset button.

mkresin commented on target/linux/ramips/image/mt76x8.mk:
```diff
@@ -299,3 +299,11 @@ define Device/zbtlink_zbt-we1226
+define Device/b-link_bl-r7628ae4
+  DTS := BL-R7628AE4
+  BLOCKSIZE := 64k
+  IMAGE_SIZE := 3776k
```
please use `IMAGE_SIZE := $(ralink_default_fw_size_4M)` here.

mkresin reviewed (changes requested):
Excellent job, looks good so far! I only found some minor issues.

Please use ```git commit --amend``` to add changes to your commit. Use ```git push --force``` to overwrite your github repository/branch. The PR will afterwards update automatically. Please don't close this PR and open a new one.

mkresin commented:
The commit message mentions seven LEDs but only one is added in the dts. What's up with the remaining leds?

Are the LAN1-4, WAN leds are working out of the box? It might be the power LED might is directly connected to VCC and can't be controlled. If you've a look at the manual of the stock firmware, some power led blinking pattern might be mentioned.

ysc3839 commented:
@mkresin WAN and LAN1-4 is controlled by hardware. Power LED can't be controlled. I didn't see any blink of power led when using stock firmware. And there's no option to turn it off.
WAN/LAN led is always blinking when the port is connected. I found a way to fix it and opened another pr (#911).

mkresin commented:
> @mkresin WAN and LAN1-4 is controlled by hardware. Power LED can't be controlled. I didn't see any blink of power led when using stock firmware.

Please add this information to the commit message.

> And there's no option to turn it off.

Should be possible by switching the pins to GPIO mode.

```
ralink,group = "p4led_an", "p3led_an", "p2led_an", "p1led_an", "p0led_an", "p4led_kn", "p3led_kn", "p2led_kn", "p1led_kn", "p0led_kn";
```

The *an pins are `&gpio1 7` to `&gpio1 11`, *kn pins are `&gpio1 30`, `&gpio1 31`, `&gpio2 0`, `&gpio2 1` and `&gpio2 3`.

Some users prefer to have full control over the LEDs and do the blinking in software using the led_switch trigger. Other prefer to have a hardware controlled blinking to not unnecessary waste CPU cycles. In the end it is up to the contributor.

Please test any changes, before updating the PR. You last update was a bit to fast :-)

ysc3839 commented:
The dts I use for test is here: https://gist.github.com/ysc3839/1bfa930f4a29e0eddaad142149ab48aa

mkresin commented:
> But when I set "switch0" led trigger and correct "port_mask", the led doesn't blink, it just light up.

Sounds like a missing feature of the switch driver. A `get_port_stats()` function is required, as it was added for other switch drivers with 0369e358916ef092a1644334f5dd1412051b68a4, 3056d09b4046e0eb0f6de0f3f5432cd9fa86fc51, 4ddbc43cc15c2fa128a2f169964ef7eb508cf2c5, 4d8a66d9346373c2a7fcac5bdae3f662a9dbd9df and 2b3ab0208eddadbf4413576eb57e608ee6e45d5c.

mkresin commented:
> All of them are mt7628. Maybe they have same led issue and must be controlled by software?

Most of the boards have a single LED to indicate the LAN status. Means, they are using already software based blinking/link indication.

The TL-WR841NV13 has all lan leds set to gpio and uses the switch led trigger. But I wouldn't be surprised if more boards are broken.

Might be that more people had the issue and you are the first who tries to fix it.

ysc3839 commented:
@mkresin 
```
root@OpenWrt:/sys/devices/platform/gpio-leds/leds/bl-r7628ae4:green:p3led_an# cat mode
link tx rx
```

mkresin commented:
Should this PR merged even if the LEDs show this strange blinking pattern (#911 not merged)?

ysc3839 commented:
@mkresin Yes.

mkresin commented:
Merged to my [staging tree](https://git.openwrt.org/?p=openwrt/staging/mkresin.git;a=summary). Please build and test an image from my staging tree, to ensure it still works as intended. Please report your results here, I'll only push the PR to master after getting the confirmation that it still works.

ysc3839 commented:
@mkresin Sorry for my late. I'm busy these days. I'll try to report it as soon as possible.

ysc3839 commented:
@mkresin Found a new bug: overlayfs is /tmp/root.
```
root@OpenWrt:~# mount
/dev/root on /rom type squashfs (ro,relatime)
proc on /proc type proc (rw,nosuid,nodev,noexec,noatime)
sysfs on /sys type sysfs (rw,nosuid,nodev,noexec,noatime)
tmpfs on /tmp type tmpfs (rw,nosuid,nodev,noatime)
tmpfs on /tmp/root type tmpfs (rw,noatime,mode=755)
overlayfs:/tmp/root on / type overlay (rw,noatime,lowerdir=/,upperdir=/tmp/root/upper,workdir=/tmp/root/work)
tmpfs on /dev type tmpfs (rw,nosuid,relatime,size=512k,mode=755)
devpts on /dev/pts type devpts (rw,nosuid,noexec,relatime,mode=600,ptmxmode=000)
debugfs on /sys/kernel/debug type debugfs (rw,noatime)
```
`lede-17.01` branch do not have this issue. So I think it's not my fault.

ysc3839 commented:
@mkresin
```
[04010C0D][04010C0C]

DDR Calibration DQS reg = 00008687



U-Boot 1.1.3 (Aug 25 2016 - 11:14:12)


Board: Ralink APSoC DRAM:  32 MB

relocate_code Pointer at: 81fb8000

******************************

Software System Reset Occurred

******************************

flash manufacture id: c2, device id 20 16

find flash: MX25L3205D

*** Warning - bad CRC, using default environment


============================================ 

Ralink UBoot Version: 4.3.0.0

-------------------------------------------- 

ASIC 7628_MP (Port5<->None)

DRAM component: 256 Mbits DDR, width 16

DRAM bus: 16 bit

Total memory: 32 MBytes

Flash component: SPI Flash

Date:Aug 25 2016  Time:11:14:12

============================================ 

icache: sets:512, ways:4, linesz:32 ,total:65536

dcache: sets:256, ways:4, linesz:32 ,total:32768 


 ##### The CPU freq = 575 MHZ #### 

 estimate memory size =32 Mbytes

RESET MT7628 PHY!!!!!!

Please choose the operation: 

   1: Load system code to SDRAM via TFTP. 

   2: Load system code then write to Flash via TFTP. 

   3: Boot system code via Flash (default).

   4: Entr boot command line interface.

   7: Load Boot Loader code then write to Flash via Serial. 

   9: Load Boot Loader code then write to Flash via TFTP. 

default: 3

>>>Chip type is MT7628A, GPIO is 6<<<



You choosed 3


 0 

   

3: System Boot system code via Flash.

## Booting image at bc050000 ...

   Image Name:   MIPS OpenWrt Linux-4.14.37

   Image Type:   MIPS Linux Kernel Image (lzma compressed)

   Data Size:    1454021 Bytes =  1.4 MB

   Load Address: 80000000

   Entry Point:  80000000

   Verifying Checksum ... OK

   Uncompressing Kernel Image ... OK

No initrd

## Transferring control to Linux (at address 80000000) ...

## Giving linux memsize in MB, 32


Starting kernel ...


[    0.000000] Linux version 4.14.37 (richard@ubuntu-vm) (gcc version 7.3.0 (OpenWrt GCC 7.3.0 r6846-b95840205d)) #0 Sat May 5 06:49:46 2018
[    0.000000] Board has DDR1
[    0.000000] Analog PMU set to hw control
[    0.000000] Digital PMU set to hw control
[    0.000000] SoC Type: MediaTek MT7628AN ver:1 eco:2
[    0.000000] bootconsole [early0] enabled
[    0.000000] CPU0 revision is: 00019655 (MIPS 24KEc)
[    0.000000] MIPS: machine is B-Link BL-R7628AE4
[    0.000000] Determined physical RAM map:
[    0.000000]  memory: 02000000 @ 00000000 (usable)
[    0.000000] Initrd not found or empty - disabling initrd
[    0.000000] Primary instruction cache 64kB, VIPT, 4-way, linesize 32 bytes.
[    0.000000] Primary data cache 32kB, 4-way, PIPT, no aliases, linesize 32 bytes
[    0.000000] Zone ranges:
[    0.000000]   Normal   [mem 0x0000000000000000-0x0000000001ffffff]
[    0.000000] Movable zone start for each node
[    0.000000] Early memory node ranges
[    0.000000]   node   0: [mem 0x0000000000000000-0x0000000001ffffff]
[    0.000000] Initmem setup node 0 [mem 0x0000000000000000-0x0000000001ffffff]
[    0.000000] random: get_random_bytes called from start_kernel+0x8c/0x474 with crng_init=0
[    0.000000] Built 1 zonelists, mobility grouping on.  Total pages: 8128
[    0.000000] Kernel command line: console=ttyS0,57600 rootfstype=squashfs,jffs2
[    0.000000] PID hash table entries: 128 (order: -3, 512 bytes)
[    0.000000] Dentry cache hash table entries: 4096 (order: 2, 16384 bytes)
[    0.000000] Inode-cache hash table entries: 2048 (order: 1, 8192 bytes)
[    0.000000] Writing ErrCtl register=00046803
[    0.000000] Readback ErrCtl register=00046803
[    0.000000] Memory: 27632K/32768K available (3352K kernel code, 167K rwdata, 812K rodata, 204K init, 208K bss, 5136K reserved, 0K cma-reserved)
[    0.000000] SLUB: HWalign=32, Order=0-3, MinObjects=0, CPUs=1, Nodes=1
[    0.000000] NR_IRQS: 256
[    0.000000] intc: using register map from devicetree
[    0.000000] CPU Clock: 575MHz
[    0.000000] timer_probe: no matching timers found
[    0.000000] clocksource: MIPS: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 6647862422 ns
[    0.000011] sched_clock: 32 bits at 287MHz, resolution 3ns, wraps every 7469508094ns
[    0.015367] Calibrating delay loop... 380.92 BogoMIPS (lpj=1904640)
[    0.087541] pid_max: default: 32768 minimum: 301
[    0.096892] Mount-cache hash table entries: 1024 (order: 0, 4096 bytes)
[    0.109849] Mountpoint-cache hash table entries: 1024 (order: 0, 4096 bytes)
[    0.129166] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 19112604462750000 ns
[    0.148537] futex hash table entries: 256 (order: -1, 3072 bytes)
[    0.160628] pinctrl core: initialized pinctrl subsystem
[    0.171407] NET: Registered protocol family 16
[    0.201711] mt7621_gpio 10000600.gpio: registering 32 gpios
[    0.212916] mt7621_gpio 10000600.gpio: registering 32 gpios
[    0.224029] mt7621_gpio 10000600.gpio: registering 32 gpios
[    0.238731] clocksource: Switched to clocksource MIPS
[    0.249821] NET: Registered protocol family 2
[    0.259216] TCP established hash table entries: 1024 (order: 0, 4096 bytes)
[    0.272904] TCP bind hash table entries: 1024 (order: 0, 4096 bytes)
[    0.285427] TCP: Hash tables configured (established 1024 bind 1024)
[    0.298123] UDP hash table entries: 256 (order: 0, 4096 bytes)
[    0.309578] UDP-Lite hash table entries: 256 (order: 0, 4096 bytes)
[    0.322195] NET: Registered protocol family 1
[    0.333613] Crashlog allocated RAM at address 0x1f00000
[    0.345123] workingset: timestamp_bits=30 max_order=13 bucket_order=0
[    0.363466] squashfs: version 4.0 (2009/01/31) Phillip Lougher
[    0.374900] jffs2: version 2.2 (NAND) (SUMMARY) (LZMA) (RTIME) (CMODE_PRIORITY) (c) 2001-2006 Red Hat, Inc.
[    0.402350] io scheduler noop registered
[    0.410015] io scheduler deadline registered (default)
[    0.420963] Serial: 8250/16550 driver, 3 ports, IRQ sharing disabled
[    0.434597] console [ttyS0] disabled
[    0.441613] 10000c00.uartlite: ttyS0 at MMIO 0x10000c00 (irq = 28, base_baud = 2500000) is a 16550A
[    0.459444] console [ttyS0] enabled
[    0.459444] console [ttyS0] enabled
[    0.473177] bootconsole [early0] disabled
[    0.473177] bootconsole [early0] disabled
[    0.489505] cacheinfo: Failed to find cpu0 device node
[    0.499708] cacheinfo: Unable to detect cache hierarchy for CPU 0
[    0.512482] spi-mt7621 10000b00.spi: sys_freq: 191666666
[    0.527004] m25p80 spi0.0: mx25l3205d (4096 Kbytes)
[    0.536771] 5 fixed-partitions partitions found on MTD device spi0.0
[    0.549354] Creating 5 MTD partitions on "spi0.0":
[    0.558852] 0x000000000000-0x000000020000 : "u-boot"
[    0.569600] 0x000000020000-0x000000030000 : "blink"
[    0.580118] 0x000000030000-0x000000040000 : "config"
[    0.590831] 0x000000040000-0x000000050000 : "factory"
[    0.601724] 0x000000050000-0x000000400000 : "firmware"
[    0.631752] 2 uimage-fw partitions found on MTD device firmware
[    0.643518] 0x000000050000-0x0000001b3005 : "kernel"
[    0.654236] 0x0000001b3005-0x000000400000 : "rootfs"
[    0.664911] mtd: device 6 (rootfs) set to be root filesystem
[    0.677395] 1 squashfs-split partitions found on MTD device rootfs
[    0.689710] 0x0000003e0000-0x000000400000 : "rootfs_data"
[    0.702042] libphy: Fixed MDIO Bus: probed
[    0.722617] rt3050-esw 10110000.esw: link changed 0x00
[    0.734357] mtk_soc_eth 10100000.ethernet eth0: mediatek frame engine at 0xb0100000, irq 5
[    0.752537] NET: Registered protocol family 10
[    0.764952] Segment Routing with IPv6
[    0.772387] NET: Registered protocol family 17
[    0.781242] 8021q: 802.1Q VLAN Support v1.8
[    0.802412] VFS: Mounted root (squashfs filesystem) readonly on device 31:6.
[    0.817533] Freeing unused kernel memory: 204K
[    0.826357] This architecture does not have kernel memory protection.
[    1.926814] init: Console is alive
[    1.933856] init: - watchdog -
[    2.368737] random: fast init done
[    3.053260] kmodloader: loading kernel modules from /etc/modules-boot.d/*
[    3.263766] kmodloader: done loading kernel modules from /etc/modules-boot.d/*
[    3.288613] init: - preinit -
[    4.945320] rt3050-esw 10110000.esw: link changed 0x00
Press the [f] key and hit [enter] to enter failsafe mode
Press the [1], [2], [3] or [4] key and hit [enter] to select the debug level
[    8.276251] mount_root: jffs2 not ready yet, using temporary tmpfs overlay
[    8.314324] urandom-seed: Seed file not found (/etc/urandom.seed)
[    8.439863] procd: - early -
[    8.445660] procd: - watchdog -
[    9.088101] procd: - watchdog -
[    9.094652] procd: - ubus -
[    9.278077] random: ubusd: uninitialized urandom read (4 bytes read)
[    9.291697] random: ubusd: uninitialized urandom read (4 bytes read)
[    9.304875] random: ubusd: uninitialized urandom read (4 bytes read)
[    9.317559] random: ubusd: uninitialized urandom read (4 bytes read)
[    9.330426] random: ubusd: uninitialized urandom read (4 bytes read)
[    9.343074] random: ubusd: uninitialized urandom read (4 bytes read)
[    9.355833] random: ubusd: uninitialized urandom read (4 bytes read)
[    9.368609] random: ubusd: uninitialized urandom read (4 bytes read)
[    9.381618] procd: - init -
Please press Enter to activate this console.
[    9.877410] kmodloader: loading kernel modules from /etc/modules.d/*
[    9.894733] ip6_tables: (C) 2000-2006 Netfilter Core Team
[    9.917129] Loading modules backported from Linux version wt-2017-11-01-0-gfe248fc2c180
[    9.933045] Backport generated by backports.git v4.14-rc2-1-31-g86cf0e5d
[    9.949511] ip_tables: (C) 2000-2006 Netfilter Core Team
[    9.968888] nf_conntrack version 0.5.0 (1024 buckets, 4096 max)
[   10.037570] xt_time: kernel timezone is -0000
[   10.100700] mt76_wmac 10300000.wmac: ASIC revision: 76280001
[   10.116627] mt76_wmac 10300000.wmac: Firmware Version: _e2_mp
[   10.128079] mt76_wmac 10300000.wmac: Build Time: 20150211175503
[   10.158724] firmware init done
[   10.437883] PPP generic driver version 2.4.2
[   10.449122] NET: Registered protocol family 24
[   10.462106] kmodloader: done loading kernel modules from /etc/modules.d/*
[   11.733077] random: jshn: uninitialized urandom read (4 bytes read)
[   11.891310] random: jshn: uninitialized urandom read (4 bytes read)
[   20.109937] jffs2: Too few erase blocks (2)
[   22.981847] br-lan: port 1(eth0.1) entered blocking state
[   22.992608] br-lan: port 1(eth0.1) entered disabled state
[   23.003809] device eth0.1 entered promiscuous mode
[   23.013353] device eth0 entered promiscuous mode
[   23.130238] br-lan: port 1(eth0.1) entered blocking state
[   23.140988] br-lan: port 1(eth0.1) entered forwarding state
[   23.152223] IPv6: ADDRCONF(NETDEV_UP): br-lan: link is not ready
[   23.978986] IPv6: ADDRCONF(NETDEV_CHANGE): br-lan: link becomes ready
```

mkresin commented:
Exactly like I suspected (you might want to check the bootlog for errors before claiming a bug/breakage).

> [    0.689710] 0x0000003e0000-0x000000400000 : "rootfs_data"
> [   20.109937] jffs2: Too few erase blocks (2)

You have 0x20000 bytes left for the rootfs_data partition. The erase block size of your board is 64Kbyte (0x10000) and at least 3 erase blocks are required for the jffs2 overlay.

There are some ways to squeeze even the last bits out of the flash. One way would be to remove unnecessary packages like ppp. But that has do be done in a custom build. Another way would be to enable 4K erase blocks.

The mx25l3205d flash chip is able to handle 4KByte erase blocks. But `CONFIG_MTD_SPI_NOR_USE_4K_SECTORS=y` isn't enabled in `target/linux/ramips/mt76x8/config-4.14`.

If you add a patch to enable CONFIG_MTD_SPI_NOR_USE_4K_SECTORS, please add `CONFIG_MTD_SPI_NOR_USE_4K_SECTORS_LIMIT=4096` as well, to ensure 4K support is only used for 4MByte flash chips. Enabling 4K sectors on supported chips, slows down writing/erasing massively.

What ever you do (or don't do), you should be aware that it might be that OpenWrt will not fit on 4MByte flash chips with the next kernel update.

ysc3839 commented:
So close this PR. Since it's flash is too small and have no external ports (USB, SD card), I think nobody will buy this router and flash OpenWrt.
Thanks for your help!
